MAGN-9861: Add path to icons, default values, search tags, tooltips of port in Library XML

aosyatnik opened the pull request:
### Purpose

Link to YouTrack:
[MAGN-9861](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-9861) Dynamo Dictionary: Add path to icons, default values, search tags, and tooltips of port Library XML snapshot

That's how XML looks now (Color.ByRGB node):
```xml
<Dynamo.Search.SearchElements.ZeroTouchSearchElement>
        <FullCategoryName>Core.Color</FullCategoryName>
        <Name>ByARGB</Name>
        <Group>Create</Group>
        <Description>Construct a color by alpha, red, green, and blue components.</Description>
        <SearchTags>color</SearchTags>
        <Inputs>
          <InputParameter Name="a" Type="int" DefaultValue="255" Tooltip="The alpha value.&#xA;&#xA;int&#xA;Default value : 255" />
          <InputParameter Name="g" Type="int" DefaultValue="0" Tooltip="The green value.&#xA;&#xA;int&#xA;Default value : 0" />
        </Inputs>
        <Outputs>
          <OutputParameter Type="Color" Tooltip="Color." />
        </Outputs>
        <SmallIcon>D:\Projects\Dynamo\Dynamo_WorkingPlace\Dynamo\bin\AnyCPU\Debug\..\..\..\src\Resources\DSCoreNodes\SmallIcons\DSCore.Color.ByARGB.Small.png</SmallIcon>
        <LargeIcon>D:\Projects\Dynamo\Dynamo_WorkingPlace\Dynamo\bin\AnyCPU\Debug\..\..\..\src\Resources\DSCoreNodes\LargeIcons\DSCore.Color.ByARGB.Large.png</LargeIcon>
</Dynamo.Search.SearchElements.ZeroTouchSearchElement>
```

I had to rename some icons and folders, because after namespace changes we forgot to do it.
Also there are new nodes and we don't have icons for them:
* Autodesk.DesignScript.Geometry.CoEdge.Previous
* Autodesk.DesignScript.Geometry.CoEdge.Next
* Autodesk.DesignScript.Geometry.CoEdge.Partner
* Autodesk.DesignScript.Geometry.CoEdge.Edge
* Autodesk.DesignScript.Geometry.CoEdge.Loop
* Autodesk.DesignScript.Geometry.CoEdge.StartVertex
* Autodesk.DesignScript.Geometry.CoEdge.EndVertex
* Autodesk.DesignScript.Geometry.CoEdge.Reversed
* Autodesk.DesignScript.Geometry.Edge.CoEdges
* Autodesk.DesignScript.Geometry.Loop.Face
* Autodesk.DesignScript.Geometry.Loop.CoEdges
* Autodesk.DesignScript.Geometry.Face.Loops
* Autodesk.DesignScript.Geometry.EllipseArc.Plane
* Analysis.Label.ByPointAndString
* Analysis.AnalysisExtensions.IsAlmostEqualTo
* Display.Display.ByGeometryColor
* Display.Display.BySurfaceColors
* DSCore.Math.Log10 - just Large

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.

### Reviewers

@pbidenko 
@marimano 

### FYIs

@Racel

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,110 @@ internal XmlDocument ComposeXmlForLibrary()
+
+                    // Case for UI nodes as ColorRange.
+                    if (dynamoNode.InPorts.Count == entry.InputParameters.Count()
+                        && entry.InputParameters.First().Item2 != Properties.Resources.NoneString)
```
what is the logic for always checking the first InputParameter - shouldn't we be checking the InputParameter at i?

aosyatnik replied:
No, here everything is written correct. We have little problem with UI nodes.
Some of them do not have incoming ports in `NodeSearchElement`, but do have incoming ports in `NodeModel`.
E.g. node List.Create.
List initializes its incoming ports, when constructor is called. So, here we check if nodeModel has input ports and nodeSearchElement also has the same input ports.
In case of list.create node it will have several incoming ports on nodeModel, but 0 incoming ports on nodeSearchElement.

mjkkirschner replied:
@aosyatnik can you please add a brief comment here explaining exactly what you just told me :)

mjkkirschner commented:
@aosyatnik maybe I am missing something but in your sample above in the test of the xml that I dumped there does not seem to be enough input parameters... in the color node there should be 4 input parameters, is that correct? but theres only two?

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,110 @@ internal XmlDocument ComposeXmlForLibrary()
+                            dynamoNode.InPorts[i].DefaultValue.ToString());
+                    }
+                    XmlHelper.AddAttribute(parameterNode, "Tooltip", dynamoNode.InPorts[i].ToolTipContent);
+                    i++;
```
ahh, you are incrementing this indexer twice.

aosyatnik replied:
hmmm, didn't notice it :(

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,110 @@ internal XmlDocument ComposeXmlForLibrary()
+                    }
+
+                    XmlHelper.AddAttribute(parameterNode, "Tooltip", dynamoNode.OutPorts[i].ToolTipContent);
+                    i++;
```
same issue here

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,110 @@ internal XmlDocument ComposeXmlForLibrary()
-            if (entry.OutputParameters.First() != Properties.Resources.NoneString)
+            if (dynamoNode.OutPorts.Count != 0)
             {
                 var inputNode = XmlHelper.AddNode(element, "Outputs");
```
please rename this variable to outputNode

aosyatnik commented:
@mjkkirschner , that was incorrect, because i++. Sorry, didn't notice...

```xml
<Dynamo.Search.SearchElements.ZeroTouchSearchElement>
        <FullCategoryName>Core.Color</FullCategoryName>
        <Name>ByARGB</Name>
        <Group>Create</Group>
        <Description>Construct a color by alpha, red, green, and blue components.</Description>
        <SearchTags>color</SearchTags>
        <Inputs>
          <InputParameter Name="a" Type="int" DefaultValue="255" Tooltip="The alpha value.&#xA;&#xA;int&#xA;Default value : 255" />
          <InputParameter Name="r" Type="int" DefaultValue="0" Tooltip="The red value.&#xA;&#xA;int&#xA;Default value : 0" />
          <InputParameter Name="g" Type="int" DefaultValue="0" Tooltip="The green value.&#xA;&#xA;int&#xA;Default value : 0" />
          <InputParameter Name="b" Type="int" DefaultValue="0" Tooltip="The blue value.&#xA;&#xA;int&#xA;Default value : 0" />
        </Inputs>
        <Outputs>
          <OutputParameter Type="Color" Tooltip="Color." />
        </Outputs>
        <SmallIcon>D:\Projects\Dynamo\Dynamo_WorkingPlace\Dynamo\bin\AnyCPU\Debug\..\..\..\src\Resources\DSCoreNodes\SmallIcons\DSCore.Color.ByARGB.Small.png</SmallIcon>
        <LargeIcon>D:\Projects\Dynamo\Dynamo_WorkingPlace\Dynamo\bin\AnyCPU\Debug\..\..\..\src\Resources\DSCoreNodes\LargeIcons\DSCore.Color.ByARGB.Large.png</LargeIcon>
</Dynamo.Search.SearchElements.ZeroTouchSearchElement>
```

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,108 @@ internal XmlDocument ComposeXmlForLibrary()
-                    var parameterNode = XmlHelper.AddNode(inputNode, "OutputParameter");
-                    XmlHelper.AddAttribute(parameterNode, "Type", parameter);
+                    var parameterNode = XmlHelper.AddNode(outputNode, "OutputParameter");
+
```
why don't we add the Name for output ports?

mjkkirschner commented on src/DynamoCore/Search/NodeSearchModel.cs:
```diff
@@ -48,58 +50,108 @@ internal XmlDocument ComposeXmlForLibrary()
+                    if (dynamoNode.OutPorts.Count == entry.OutputParameters.Count()
+                        && entry.OutputParameters.First() != Properties.Resources.NoneString)
+                    {
+                        XmlHelper.AddAttribute(parameterNode, "Type", entry.OutputParameters.ElementAt(i));
```
<img width="1123" alt="screen shot 2016-04-14 at 2 42 20 pm" src="https://cloud.githubusercontent.com/assets/508936/14540367/a5f44fd4-0252-11e6-9924-2c6503fadb6d.png">

I see a lot of nodes that don't have output Type attributes - why is this? I see this for some zero touch nodes as well like color.components.

aosyatnik replied:
Color.Components out ports are set as [MultiReturn("a", "r", "g", "b")].
We didn't add types for `MultiReturn`:( Maybe, we should create additional task for this?
Even in Dynamo it's not said what type port returns.... `var` is tooltip by default.
http://prntscr.com/asi61m

aosyatnik replied:
List.Map is not zerotouch. It's UI node. It's the same issue as described before. (look comment about List.Create)

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1475,7 +1475,7 @@ internal void DumpLibraryToXml(object parameter)
             string fullFileName = Path.Combine(pathManager.LogDirectory, fileName);
 
-            SearchModel.DumpLibraryToXml(fullFileName);
+            SearchModel.DumpLibraryToXml(fullFileName, this.PathManager.DynamoCoreDirectory);
```
just want to check - have you tested this on Revit? does it work?

mjkkirschner commented:
@aosyatnik thanks for the other fixes - but I still have some other questions:

* There are a lot of nodes with Output Parameters with no Types, why not?
* The output ports don't have their names recorded - is there any reason we cant do this?
* can you please describe the issue for UI nodes in a comment above those cases in the code.
* Have you tested this code running on DynamoForRevit? does it pickup the correct path ?

aosyatnik commented:
@mjkkirschner I'll try answer questions...

* There are a lot of nodes with Output Parameters with no Types, why not?
It's because their output ports are set as `MultiReturn` attribute, which doesn't support types.

* The output ports don't have their names recorded - is there any reason we cant do this?
No, fixed.

* can you please describe the issue for UI nodes in a comment above those cases in the code.
Sure, added comment.

* Have you tested this code running on DynamoForRevit? does it pickup the correct path ?
Yes. `PathManager.DynamoCoreDirectory` is the same as in DynamoSandbox.
BUT: icons for Revit specific nodes are in Revit repo and in `DynamoRevit\src\Resources` folder... Should we move all icons in the one folder `Dynamo\src\Resources`?

One more issue, these icons folder can be tested just with DynamoSandbox build. I mean our testers won't be able test it with installer, because icons are not part of installer.

ramramps commented:
@aosyatnik - This task to get the XML. So, do not move the icons.
For the last question, I assume the XML is good on DynamoRevit. True?

aosyatnik commented:
Replaced by https://github.com/DynamoDS/Dynamo/pull/6463

aosyatnik commented:
Hi, @ramramps , I created new PR just with code changes. 

For the last question, I assume the XML is good on DynamoRevit. True?
No.  Icons for Revit specific nodes are in Revit repository, in DynamoRevit\src\Resources folder.
All other icons are in Dynamo repository, in Dynamo\src\Resources folder.

To work correct icons should be moved in one folder.
